nvaccess/nvda
# Text alignment reporting and enhancements for MS Excel cells and MS Word text

### Link to issue number:
Closes #15206
Closes #15220

### Summary of the issue:

Various issues for alignment reporting:
* In MS Excel horizontal or vertical text alignment is not always correctly reported.
* In MS Word (legacy), text alignment is not correctly reported for "justify" and for "distrubute".

### Description of user facing changes
Reporting text alignment in Word and Excel is now handled better in various situations. "Default" is not reported anymore since it does not give any clear indication of the position of the text.

* In MS Excel (no UIA) the alignment is reported as per Excel's options:
  * for horizontal text alignment in cell:  "General", "Centered across columns", "Repeat" and "Distribute" is now reported instead of "Default" or nothing
  * for vertical alignment: "bottom" is reported instead of "default"
* In MS Word (no UIA):
  * "justify" is now reported instead of "Default"
  * "distribute" is now reported instead of nothing

### Description of development approach

Standardized alignments returned by `getFormatField*` functions around a string enum as done in #11897. Thus, all the interfaces are impacted.

### Testing strategy:
Test plan based on #11897, adapted to this PR.

For the following APIs/files, the following test are made:
* Check that NVDA+F is able to report text alignment
* Activate the option to report text alignment and navigate with the caret to test the transitions between the various possible text alignments.

API/files tested:
- [x] NVDAObjects/IAccessible/__init__.py:
      * Test with Chrome: `data:text/html,<p align="left">left</p><p align="right">right</p><p align="center">center</p><p align="justify">justify</p>`
        OK for left, right, center, justify and match-parent; NOK for experimental justify-all, start, end without or with this PR
      * Edge (legacy)
        OK for left, right, center, justify; NOK for experimental justify-all, start, end without or with this PR
- [x] NVDAObjects/UIA/__init__.py: test with Word (UIA); left, right, center, justify are supported; but distribute is reported as justify
- [x] NVDAObjects/window/edit.py, class ITextDocumentTextInfo: Test with Wordpad (left, right, center, justify)
- [x] NVDAObjects/window/edit.py, class EditTextInfo: Paragraph alignment is not supported by NVDA (probably not natively either; no test done.
- [x]  NVDAObjects/window/excel.py: test with Excel, cell navigation (alignment of full cell).
  * The following horizontal text alignments are reported (these are the ones available in cell formatting dialog): general, left, right, center, centered across multiple columns, justify, fill and distributed.
  * The following vertical text alignments are reported (these are the ones available in cell formatting dialog): top, centered, bottom, justitfy and distributed.
  * Note: When editing the cell's content, no formatting can be reported (neither before this PR nor with this PR).
- [x] NVDAObjects/window/winword.py: test with Word 2016 (UIA not checked): left, right, center, justify and distribute are reported
- [x] virtualBuffers/MSHTML.py: Test `ui.browseableMessage`:
      * Test case:
      `ui.browseableMessage('''<p align="left">left</p><p align="right">right</p><p align="center">center</p><p align="justify">justify</p><p align="justify-all">justify-all</p><p align="start">start</p><p align="end">end</p><div align="center"><p align="match-parent">match-parent center</p></div>''', "Test", isHtml=True)`
      * Result: no change with respect to NVDA 2023.2beta1, i.e. OK for left, right, center, justify; NOK for start, end and justify-all which are all reported as left.


Tests done with:
* MS Office 2016 (16.0.16529.20182).
* Chrome 115.0.5790.110
* Edge 114.0.1823.82


#### Special case of Firefox

A part of the work of this PR was done some time ago and was working with Firefox. Today the situation is the following:
* Reporting text alignment in Firefox 116.0 (installed) is not supported, neither in NVDA 2023.2beta1 nor with this PR
* With Firefox ESR portable 102.13.0 (portable) and NVDA 2023.2beta1, text alignment is reported for justified text, but "default" is reported for left, right and centered text.
* With Firefox ESR portable 102.13.0 (portable) and this PR, text alignment is correctly reported for left, right, center and justify.

I imagine that there may be an issue with latest Firefox versions regarding text alignment reporting. @jcsteh any information or confirmation of this?

### Known issues with pull request:

1. With Word no UIA:
Not all the available alignments have been implemented. The following are missing: wdAlignParagraphJustifyHi, wdAlignParagraphJustifyLow, wdAlignParagraphJustifyMed, wdAlignParagraphThaiJustify. They were not supported before by NVDA and do not know how to use them; maybe with some specific locale?
(see https://docs.microsoft.com/en-us/office/vba/api/word.wdparagraphalignment)
If there is a need to implement them in the future, it should be quite easy.

2. Word with UIA: "distribute" is reported as "justify", since "distribute" does not seem to be a supported alignment in UIA.

3. In Excel with UIA, text alignment is not supported at all today (in NVDA 2023.2beta1). This PR does not add the support for text alignment in Excel with UIA.

4. Text alignment reporting is not supported by NVDA in:
   * PowerPoint (see #4021)
   * Libre Office Writer (see #2204)
   * Edge with UIA
 
 This PR does not improve it.

5. For web text alignment, according to [this page](https://w3c.github.io/i18n-tests/results/text-align), `text-align="justify-all"` is not supported by any browser. NVDA reports it as "left" before and after this PR.

### Change log entries:
Bug fixes
`In Word and Excel, text alignment will be correctly reported in more situations. (#15206, #15220)`

### Code Review Checklist:

  - description is up to date
  - change log entries
  - Unit tests
  - System (end to end) tests
  - Manual testing
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
